src/overlay/SurveyDataManager.h

```diff
@@ -0,0 +1,206 @@
+#pragma once
+
+// Copyright 2024 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "main/Config.h"
+#include "overlay/Peer.h"
+#include "util/NonCopyable.h"
+#include "util/Timer.h"
+#include "util/UnorderedMap.h"
+
+#include "medida/meter.h"
+#include "xdr/Stellar-overlay.h"
+#include "xdr/Stellar-types.h"
+
+#include <chrono>
+#include <mutex>
+#include <optional>
+
+namespace stellar
+{
+
+enum class SurveyPhase
+{
+    // Survey is currently recording data
+    RECORDING,
+    // Recording complete. Survey data is available for reporting.
+    REPORTING,
+    // No active survey in progress. No data is being recording or reported.
+    INACTIVE
+};
+
+struct RecordingNodeData
+{
+    RecordingNodeData(uint64_t initialLostSyncCount);
+
+    // Peer change data
+    uint32_t mAddedAuthenticatedPeers = 0;
+    uint32_t mDroppedAuthenticatedPeers = 0;
+
+    // SCP stats
+    std::vector<std::chrono::nanoseconds> mSCPFirstToSelfLatencies;
+    std::vector<std::chrono::nanoseconds> mSCPSelfToOtherLatencies;
+
+    // To compute how many times the node lost sync in the time slice
+    uint64_t const mInitialLostSyncCount;
+};
+
+// Data about a peer
+struct RecordingPeerData
+{
+    RecordingPeerData(Peer::PeerMetrics const& peerMetrics);
+
+    // Metrics at the start of the survey
+    uint64_t const mInitialMessageRead;
+    uint64_t const mInitialMessageWrite;
+    uint64_t const mInitialByteRead;
+    uint64_t const mInitialByteWrite;
+    uint64_t const mInitialUniqueFloodBytesRecv;
+    uint64_t const mInitialDuplicateFloodBytesRecv;
+    uint64_t const mInitialUniqueFetchBytesRecv;
+    uint64_t const mInitialDuplicateFetchBytesRecv;
+    uint64_t const mInitialUniqueFloodMessageRecv;
+    uint64_t const mInitialDuplicateFloodMessageRecv;
+    uint64_t const mInitialUniqueFetchMessageRecv;
+    uint64_t const mInitialDuplicateFetchMessageRecv;
+
+    // For computing average latency (in miliseconds)
+    std::chrono::milliseconds mLatencySum{0};
```

Comment:
No need to manually compute latency here (we can use medida)

Reply:
Done in 21fb55214f5416cf47f3d79b8a39b6caddac48cb